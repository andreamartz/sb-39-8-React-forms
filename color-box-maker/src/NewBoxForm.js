/**
 * NewBoxForm component:
 * 
 *  * renders a form that, when submitted, creates a new Box with width, height, and background color specified
 *  * When the form is submitted, clear the input values. (This requires state)
 */

import React, { useState } from "react";


const NewBoxForm = ({ addBox }) => {
  const INITIAL_STATE = {
    width: '',
    height: '',
    backgroundColor: ''
  };
  const [formData, setFormData] = useState(INITIAL_STATE);

  // handleChange will handle changes to form input fields
  const handleChange = (evt) => {
    const { name, value } = evt.target;  // name and value come from the target input
    
    setFormData(data => ({
      ...data,         // use spread operator to insert the formData we currently have
      [name]: value    // in the current formData, change only the value corresponding to the computed value of name
    }));
  }

  const handleSubmit = (evt) => {
    evt.preventDefault();
    const { width, height, backgroundColor } = formData;
    // add the new box to state
    addBox({ ...formData });
    console.log(`Created box with ${width}px width, ${height}px height, and background color ${backgroundColor}`)
    setFormData(INITIAL_STATE);
  }

  return (
    <form onSubmit={handleSubmit}>
      <label htmlFor="width">Width (px)</label>
      <input 
        id="width" 
        type="text" 
        name="width"      // name matches what is in state
        placeholder="width" 
        value={formData.width} 
        onChange={handleChange}
      >
      </input>

      <label htmlFor="height">Height (px)</label>
      <input 
        id="height" 
        type="text" 
        name="height"      // name matches what is in state
        placeholder="height" 
        value={formData.height} 
        onChange={handleChange}
      >
      </input>

      <label htmlFor="backgroundColor">Background Color</label>
      <input 
        id="backgroundColor" 
        type="text" 
        name="backgroundColor"      // name matches what is in state
        placeholder="background color" 
        value={formData.backgroundColor} 
        onChange={handleChange}
      >
      </input>
      <button>Make box</button>
    </form>
  )
}

export default NewBoxForm;