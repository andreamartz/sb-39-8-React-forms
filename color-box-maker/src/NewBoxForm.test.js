import React from "react";
import { render } from '@testing-library/react';
import NewBoxForm from './NewBoxForm';

describe("Smoke tests", () => {
  test('renders without crashing', () => {
    render(<NewBoxForm />);
  });
});

describe("Snapshot tests", () => {
  test('matches snapshot', () => {
    const {asFragment} = render(<NewBoxForm />);
    expect(asFragment()).toMatchSnapshot();
  });
});