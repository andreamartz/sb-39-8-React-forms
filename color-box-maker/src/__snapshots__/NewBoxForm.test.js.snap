// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot tests matches snapshot 1`] = `
<DocumentFragment>
  <form>
    <label
      for="width"
    >
      Width (px)
    </label>
    <input
      id="width"
      name="width"
      placeholder="width"
      type="text"
      value=""
    />
    <label
      for="height"
    >
      Height (px)
    </label>
    <input
      id="height"
      name="height"
      placeholder="height"
      type="text"
      value=""
    />
    <label
      for="backgroundColor"
    >
      Background Color
    </label>
    <input
      id="backgroundColor"
      name="backgroundColor"
      placeholder="background color"
      type="text"
      value=""
    />
    <button>
      Make box
    </button>
  </form>
</DocumentFragment>
`;
